api/src/services/export/dhis.js

```diff
@@ -0,0 +1,215 @@
+const _ = require('lodash');
+const moment = require('moment');
+
+const db = require('../../db');
+const settings = require('../settings');
+
+/**
+ * @param {string} filters.dataSet
+ * @param {integer} filters.from
+ * @param {string=} filters.orgUnit
+ *
+ * @param options.humanReadable
+ */
+module.exports = async (filters, options = {}) => {
+  const { dataSet, orgUnit } = filters;
+  const { from } = filters.date || {};
+  if (!dataSet) {
+    throw { code: 422, message: 'filter "dataSet" is required' };
+  }
+
+  if (!from) {
+    throw { code: 422, message: 'filter "from" is required' };
+  }
+
+  const settingsDoc = await settings.get();
+  const dataSetConfig = settingsDoc.dhisDataSets &&
+    Array.isArray(settingsDoc.dhisDataSets) &&
+    settingsDoc.dhisDataSets.find(dhisDataSet => dhisDataSet.guid === dataSet);
+  if (!dataSetConfig) {
+    throw { code: 422, message: `dataSet "${dataSet}" is not defined` };
+  }
+  
+  const dhisTargetDefinitions = getDhisTargetDefinitions(dataSet, settingsDoc);
+  if (dhisTargetDefinitions.length === 0) {
+    throw { code: 422, message: `dataSet "${dataSet}" has no dataElements` };
+  }
+  
+  const targetDocsAtInterval = await fetch.targetDocsAtInterval(from);
+  const contactsWithOrgUnits = await fetch.contactsWithOrgUnits(orgUnit);
+  const targetOwnerIds = _.uniq(targetDocsAtInterval.map(target => target.owner));
+  const targetOwners = await fetch.docsWithId(targetOwnerIds);
+  const mapContactIdToOrgUnit = mapContactIdToOrgUnits(dataSet, targetOwners, contactsWithOrgUnits);
+  const targetDocsInHierarchy = targetDocsAtInterval.filter(target => !orgUnit || mapContactIdToOrgUnit[target.owner]);
```

Comment:
I'm seeing a theme of a null orgUnit being OK. I'm guessing this is very intentional? Do you think this can be formalised more? In some places it looks accidental, but I'm presuming it's part of the DHIS2 spec?

Reply:
The "export" UI has the ability to export data for "all places". In this case, there is no orgUnit filter and we should export all data that we have for all configured contacts.